### PR TITLE
allow using both infrastructure_roles_options

### PR DESCRIPTION
**Pull request by FxKu**
Current master only allows to set one of the two infrastructure roles options. If `infrastructure_roles_secrets` is set, `infrastructure_roles_secret_name` is ignored. However, as the new option doesn't allow "templating", so configmap users might still want to use old option but also being able to reference another secret with the new flexibility of choosing the keys.

This PR also fixes a small "bug": When using the new configmap option one had to specify a `rolekey`. But it could be that this info isn't available. CRD validation also doesn't enforce it.

### CODE REVIEW

File: charts/postgres-operator/crds/operatorconfigurations.yaml [hunk 1]
```diff
@@ -138,7 +138,6 @@ spec:
                     type: object
                     required:
                       - secretname
-                      - userkey
```
Comment (erthalion) [inline charts/postgres-operator/crds/operatorconfigurations.yaml:141]: Why does it not required?

File: manifests/infrastructure-roles-new.yaml [hunk 2]
```diff
@@ -3,8 +3,8 @@ data:
   # infrastructure role definition in the new format
   # robot_zmon_acid_monitoring_new
   user: cm9ib3Rfem1vbl9hY2lkX21vbml0b3JpbmdfbmV3
-  # robot_zmon_new
-  role: cm9ib3Rfem1vbl9uZXc=
+  # robot_zmon
```
Comment (erthalion) [inline manifests/infrastructure-roles-new.yaml:6]: To test default values it makes sense to omit this key.

File: pkg/cluster/sync.go [hunk 3]
```diff
@@ -500,6 +500,7 @@ func (c *Cluster) syncSecrets() error {
 				c.logger.Warningf("secret %q does not contain the role %q", secretSpec.Name, secretUsername)
 				continue
 			}
+			c.Secrets[secret.UID] = secret
```
Comment (erthalion) [inline pkg/cluster/sync.go:503]: How is this related to the infrastructure roles?

File: pkg/controller/util.go [hunk 4]
```diff
@@ -118,16 +119,12 @@ var emptyName = (spec.NamespacedName{})
...
-		// All the other possibilities require secret name to be present, so if
-		// it is not, then nothing else to be done here.
-		return rolesDefs
-	}
+	// take from CRD configuration
+	rolesDefs := c.opConfig.InfrastructureRoles
 
 	// check if we can extract something from the configmap config option
-	if c.opConfig.InfrastructureRolesDefs != "" {
+	if len(rolesDefs) == 0 && c.opConfig.InfrastructureRolesDefs != "" {
```
Comment (erthalion) [inline pkg/controller/util.go:127]: No one prevents having both new roles definitions and old format, AFAICS that was the idea.

File: pkg/controller/util.go [hunk 5]
```diff
@@ -163,27 +160,33 @@ func (c *Controller) getInfrastructureRoleDefinitions() []*config.Infrastructure
...
+				roleDef.DefaultUserValue = value
+			case "defaultrolevalue":
+				roleDef.DefaultRoleValue = value
 			default:
 				c.logger.Warningf("Role description is not known: %s", properties)
 			}
 		}
-	} else {
+
+		if roleDef.SecretName != emptyName &&
```
Comment (erthalion) [inline pkg/controller/util.go:172]: It's probably better to have "validate" method.

File: pkg/controller/util_test.go [hunk 6]
```diff
@@ -301,14 +301,51 @@ func TestInfrastructureRoleDefinitions(t *testing.T) {
...
+			"secretname: infrastructureroles-new-test, userkey: test-user, passwordkey: test-password, defaultrolevalue: test-role",
+			[]*config.InfrastructureRole{
+				&config.InfrastructureRole{
+					SecretName: spec.NamespacedName{
+						Namespace: v1.NamespaceDefault,
+						Name:      testInfrastructureRolesNewSecretName,
+					},
+					UserKey:          "test-user",
+					PasswordKey:      "test-password",
+					RoleKey:          "",
```
Comment (erthalion) [inline pkg/controller/util_test.go:342]: It's probably better to test with no role value present in the secret.

File: charts/postgres-operator/crds/operatorconfigurations.yaml (hunk 1 above)
Comment (FxKu) [reply charts/postgres-operator/crds/operatorconfigurations.yaml:141]: when we can have a default user either that or the userkey has to be set. That's why I excluded it here

File: manifests/infrastructure-roles-new.yaml (hunk 2 above)
Comment (FxKu) [reply manifests/infrastructure-roles-new.yaml:6]: yes, thought about it too. Will update

File: pkg/cluster/sync.go (hunk 3 above)
Comment (FxKu) [reply pkg/cluster/sync.go:503]: moved to #1091

File: charts/postgres-operator/crds/operatorconfigurations.yaml (hunk 1 above)
Comment (erthalion) [reply charts/postgres-operator/crds/operatorconfigurations.yaml:141]: Userkey is mandatory, default value should kicks in when a userkey value is not found within a secret, no?

Comment (FxKu) [reply charts/postgres-operator/crds/operatorconfigurations.yaml]: true, mixed it up in my head

**Comment by erthalion** [comment]
:+1:

**Comment by FxKu** [comment]
:+1:
